app/models/tag_layout.rb

```diff
@@ -65,11 +72,11 @@ class TagLayout < ApplicationRecord
   delegate :walking_by, :walk_wells, :apply_tags, to: :walking_algorithm_helper
 
   def direction=(new_direction)
-    self.direction_algorithm = DIRECTION_ALGORITHMS.fetch(new_direction) { UnknownDirection.new(new_direction) }
+    self.direction_algorithm = DIRECTION_ALGORITHMS.fetch(new_direction, TagLayout::DummyDirectionModule)
```

Comment:
Im not sure why direction and walking_by must exist in validation if we are dummying their values? Maybe so other validation can occur? Im not familiar with how TagLayout works so might be a misunderstanding on my part.

Reply:
I think it's using a very backwards approach to these values, but you seem to have a similar understanding to mine.  In this case:

- `direction` is assigned a value as an attribute.
- This setter method uses the value to set `self.direction_algorithm` using the attribute value.  If that isn't a valid algorithm, the dummy value is used (you'll see why in a later step)
- The validator on line 63 expects `direction` to be present, BUT
- At this point `direction` is delegated to `direction_algorithm_module` via line 71.
- That delegation tries to get the name of the direction module (which is either a valid module, or the Dummy module from earlier) and constantise its name.
- If we didn't use a Dummy module when the value was invalid, the constantize would generate an exception.
- The `direction` method is called on that module to do the validation.  So we have a method in the Dummy module that returns `nil` which fails the validation of the model.

Convoluted.  Something similar happens for `walking_by`.

Reply:
Hmm yes rather convoluted.. I think because this story is only replicating the behaviour of the existing API the changes are fine. This does look like something that can be refactored at some point though.